### ISSUE TITLE
Test failure: System.IO.Tests.StreamReaderWithBufferedStream_Tests.ReadToEndAsync_WithCancellation

### ISSUE THREAD

**Issue by VincentBu** [dotnet/runtime#67730]
Run: [runtime-libraries-coreclr outerloop 20220407.2](https://dev.azure.com/dnceng/public/_build/results?buildId=1704798&view=ms.vss-test-web.build-test-results-tab&runId=46435126&resultId=102087&paneView=debug)

Failed test: 
```
net7.0-Linux-Release-x64-CoreCLR_release-RedHat.7.Amd64.Open

- System.IO.Tests.StreamReaderWithBufferedStream_Tests.ReadToEndAsync_WithCancellation

```
**Error message:**
```
System.IO.IOException : No space left on device : '/tmp/StreamReaderWithBufferedStream_Tests_koju4dc4.shc/ReadToEndAsync_WithCancellation_140_a25ccac0'


Stack trace
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.OSFileStreamStrategy.Write(ReadOnlySpan`1 buffer)
   at System.IO.Strategies.BufferedFileStreamStrategy.Flush(Boolean flushToDisk)
   at System.IO.Strategies.BufferedFileStreamStrategy.Dispose(Boolean disposing)
   at System.IO.StreamWriter.CloseStreamFromDispose(Boolean disposing)
   at System.IO.StreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at System.IO.Tests.StreamReaderTests.CreateLargeFile(String path) in /_/src/libraries/System.IO/tests/StreamReader/StreamReaderTests.cs:line 188
   at System.IO.Tests.StreamReaderTests.ReadToEndAsync_WithCancellation() in /_/src/libraries/System.IO/tests/StreamReader/StreamReaderTests.cs:line 141
--- End of stack trace from previous location ---
```

**Comment by adamsitnik** [comment]
> No space left on device

It's an infra issue

**Comment by danmoseley** [comment]
Area-infrastructure doesn't own the test machines.

@dotnet/dnceng  could you check on the machine used here? Is it too late to determine what is using the space?

**Comment by premun** [comment]
Unfortunately, this was running on a VM and the VM is long gone. However, the VM processed 148 work items in a row:

```sql
WorkItems
| where QueueName == "redhat.7.amd64.open.rt" and MachineName == "a00112L" and Queued >= now() - 3d
| project FriendlyName, MachineName, Queued, Finished, ExitCode, Status, PassCount, FailCount, SkipCount, ConsoleUri
```

I will have a look and check what files it was downloading. Maybe there was some huge correlation payload being downloaded.

**Comment by premun** [comment]
Looking at the payloads of the job, they don't seem overly huge (200 MB altogether downloaded to that machine).

```sql
Metrics
    | where JobName == "533002c3-65e2-406e-a3a8-d35a6d8bf9d5" and MetricName == "FileSize" and EventType == "FileDownload"
    | join kind=inner WorkItems on WorkItemId
    | where MachineName == "a00112L"
    | extend Dimensions = parse_json(Dimensions)
    | where toint(Dimensions.Cached) == 0// and tostring(Dimensions["Payload..."]) == "correlation"
    | summarize sum(MetricValue)
```

Is it possible the job creates a lot of files?

**Comment by premun** [comment]
I have also recently dealt with an FR question where we found out that some of the build happened inside of the **correlation payload directory** which is common for the whole job, is uploaded only once per job and downloaded to the machine to one place where in case it happens that an agent processes 2+ workitems from the same job, it is re-used (it is kept for some time).

Is it possible something like this happens in these jobs?

**Comment by danmoseley** [comment]
These are IO tests, so certainly it's possible that it creates a lot of files/large files. Can you think of a way that the infrastructure could help us figure out what is using the disk space? Maybe hold a VM for investigation if it his out of disk space?

**Comment by premun** [comment]
This is not a feature we currently have unfortunately. @MattGal would you advise some other debugging steps in case of problems with running out of disk space?

I am also seeing that there was probably enough disk space (47 GB) around the time this happened:
```sql
HeartbeatExport
| where Queue == "redhat.7.amd64.open.rt"
| where Machine == "a00112L"
```

I have no knowledge about the tests themselves but the stacktrace seems to include methods such as `CreateLargeFile` so I am a bit suspicious that this is an infra issue?

**Comment by MattGal** [comment]
@danmoseley  / @premun I did see one interesting thing.  Yes, this test was added a month or so ago and is designed to make big files in https://github.com/dotnet/runtime/commit/0e8422d918b859800ce6d176db0903d76ca074c0.  I'd note that it's trying to create a 2.08 GB (not 1 GB like noted in the comments) text file.

The "one interesting thing" bit is in the exception message:

'/tmp/StreamReaderWithBufferedStream_Tests_koju4dc4.shc/ReadToEndAsync_WithCancellation_140_a25ccac0'

In the helix client, we have (abridged):

``` 
        workitem_temp = os.path.join(self.settings.work_root, hashed_correlation_id, "t")
...
        os.environ["TEMP"] = workitem_temp
        os.environ["TMP"] = workitem_temp
```

So it would appear that `tempDirectory ??= Path.GetTempPath();`  is somehow escaping this (perhaps some remote executor trickery?)

If you can make it actually honor these paths, there will be more space.

### If you just want a repro machine: 

The helix-repro-vms DTL can and will definitely reproduce this problem for you if you run the test enough times, mostly a `df -ah /tmp`  should be enlightening.

**Comment by danmoseley** [comment]
@MattGal the above script will not work on Unix. 
On Unix, the variable used is TEMPDIR (https://en.wikipedia.org/wiki/TMPDIR) and my guess is that if that's not set it is hard coded to `/tmp`
Wikipedia also says "ther forms sometimes accepted are TEMP, TEMPDIR and TMP, but these alternatives are used more commonly by non-[POSIX]..."
On Windows, GetTempPath() looks at TMP first, then TEMP, then some other things that are probably undocumented. I don't see TEMPDIR used there. Maybe it has %LOCALAPPDATA%\Temp hard coded as another fallback.

Our code follows this. On Unix:
```c#
        public static string GetTempPath()
        {
            const string TempEnvVar = "TMPDIR";

            // Get the temp path from the TMPDIR environment variable.
            // If it's not set, just return the default path.
            // If it is, return it, ensuring it ends with a slash.
            string? path = Environment.GetEnvironmentVariable(TempEnvVar);
```

**Comment by danmoseley** [comment]
It looks like the `Path.GetTempPath()` code falls back to `/tmp/` if `TEMPDIR` is not defined. I suppose one could argue it maybe should check `TMP` and/or `TEMP` first, but that might be pure Windows leakage, plus would need thought about security possibly.

**Comment by MattGal** [comment]
@danmoseley so if we just also set TMPDIR that'd be OK right?  This is important for testing because we redirect not for disk space but rather per-workitem isolation.

**Comment by MattGal** [comment]
I went ahead and made this into [a PR](https://dnceng.visualstudio.com/internal/_git/dotnet-helix-machines/pullrequest/22266)

**Comment by danmoseley** [comment]
You should set TMPDIR for non Windows and TMP (or TEMP or possibly both, why not) for Windows.

If this script runs on either, I'd set all three to be sure.

**Comment by MattGal** [comment]
> You should set TMPDIR for non Windows and TMP (or TEMP or possibly both, why not) for Windows.
> 
> If this script runs on either, I'd set all three to be sure.

I went ahead and set all 4.  I also did a little experiment (test helix log [here](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/results-19ccb555290645dfa1/dumping-file-info-1/1/console.bef0ab22.log?sv=2019-07-07&se=2022-05-01T23%3A39%3A39Z&sr=c&sp=rl&sig=FJ%2B34aUjNGdk9DyxC4hpiCXSads9BGS%2FwXq5wIxbTWA%3D))

Relevant excerpt:
```
Disk Space in Human readable format
*****************************
Filesystem                 Size  Used Avail Use% Mounted on
...
/dev/mapper/rootvg-tmplv   2.0G  131M  1.9G   7% /tmp
...
/dev/sdb1                   50G   53M   47G   1% /mnt
```

This shows us /tmp has only 2 GB free with no automated cleanup, but the helix work drive (mounted in /mnt/work) gets ~47 GB of space and will automatically be cleaned regardless of what happens to your test.   So once this rolls out, disk space panic on RHEL at least will go way down.

This makes me think that I was wrong before and UTF-8 chars get encoded as 1 byte/char otherwise this could never pass.
